core/range_scan_load_balancer.cxx

```diff
@@ -0,0 +1,145 @@
+ *   limitations under the License.
+ */
+
+#include "range_scan_load_balancer.hxx"
+
+#include <algorithm>
+#include <limits>
+#include <map>
+#include <numeric>
+#include <queue>
+#include <random>
+#include <vector>
+
+namespace couchbase::core
+{
+range_scan_node_state::range_scan_node_state(std::queue<std::uint16_t> vbuckets)
+  : pending_vbuckets_{ std::move(vbuckets) }
+{
+}
+
+auto
+range_scan_node_state::fetch_vbucket_id() -> std::optional<std::uint16_t>
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    if (pending_vbuckets_.empty()) {
+        return {};
+    }
+    active_stream_count_++;
+    auto vbucket_id = pending_vbuckets_.front();
+    pending_vbuckets_.pop();
+    return vbucket_id;
+}
+
+void
+range_scan_node_state::notify_stream_ended()
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    active_stream_count_--;
+}
+
+void
+range_scan_node_state::enqueue_vbucket(std::uint16_t vbucket_id)
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    pending_vbuckets_.push(vbucket_id);
+}
+
+auto
+range_scan_node_state::active_stream_count() -> std::uint16_t
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    return active_stream_count_;
+}
+
+auto
+range_scan_node_state::pending_vbucket_count() -> std::size_t
+{
+    std::lock_guard<std::mutex> const lock{ mutex_ };
+    return pending_vbuckets_.size();
+}
+
+range_scan_load_balancer::range_scan_load_balancer(const topology::configuration::vbucket_map& vbucket_map,
+                                                   std::optional<std::uint64_t> seed)
+  : seed_{ seed }
+{
+    std::map<std::int16_t, std::queue<std::uint16_t>> node_to_vbucket_map{};
+    for (std::uint16_t vbucket_id = 0; vbucket_id < vbucket_map.size(); vbucket_id++) {
+        auto node_id = vbucket_map[vbucket_id][0];
+        if (node_to_vbucket_map.count(node_id) == 0) {
+            node_to_vbucket_map[node_id] = {};
+        }
+        node_to_vbucket_map[node_id].push(vbucket_id);
+    }
+    for (auto [node_id, vbucket_ids] : node_to_vbucket_map) {
+        nodes_.emplace(node_id, std::move(vbucket_ids));
+    }
+}
+
+void
+range_scan_load_balancer::seed(std::uint64_t seed)
+{
+    seed_ = seed;
+}
+
+auto
+range_scan_load_balancer::select_vbucket() -> std::optional<std::uint16_t>
+{
+    std::lock_guard<std::mutex> const lock{ select_vbucket_mutex_ };
+
+    auto min_stream_count = std::numeric_limits<std::uint16_t>::max();
+    std::optional<std::int16_t> selected_node_id{};
+
+    std::vector<std::map<int16_t, range_scan_node_state>::iterator> iterators{ nodes_.size() };
+
+    std::iota(iterators.begin(), iterators.end(), nodes_.begin());
+    std::mt19937_64 gen{ std::random_device{}() };
+    if (seed_.has_value()) {
+        gen.seed(seed_.value());
+    }
+    std::shuffle(iterators.begin(), iterators.end(), gen);
+
+    for (auto it : iterators) {
+        auto node_id = it->first;
+        auto* node_status = &it->second;
```

Comment:
It is sugar, but I think you can do structured binding here

```c++
        auto &[node_id, node_status] = *it;
```

Reply:
Done